### ISSUE TITLE
zipfile plugin gives error when click on another tab with mouse

### ISSUE THREAD

**Issue by user202729** [vim/vim#12136]
### Steps to reproduce

1. Open a zip file in vim.
2. Vim displays a list of files, normally if a file is clicked on then it will be opened in vim.
3. Open another tabpage, switch back to the tab that opens the zip file.
4. Click on the other tab's title in the tabline.
5. See the error

```
Error detected while processing function <SNR>88_ZipBrowseSelect:
line   24:
E121: Undefined variable: b:zipfile
line   35:
E121: Undefined variable: zipfile
E116: Invalid arguments for function fnameescape("zipfile://".zipfile.'::'.fname)
```

Seems to be caused by this line.

https://github.com/vim/vim/blob/519cc559b08b800edc429688aece7ad6a00d41eb/runtime/autoload/zip.vim#L186

Alternatively this can be reproduced by opening a new window and click on that window.

### Expected behaviour

Nothing happens.

### Version of Vim

9.0.910.

### Environment

Linux, xterm.

--------

Actually this may be more of a vim's unexpected behavior instead of zipfile's, that a `noremap <buffer> <leftmouse>` is still triggered if that mouse click switches to another buffer.

**Comment by brammool** [comment]
The mapping for LeftMouse first executes the mouse movement, thus jumps to another buffer where b:zipfile isn't set.
The solution would be that early in ZipBrowseSelect() it bails out if b:zipfile is not set.
@cecamp

**Comment by user202729** [comment]
I think even if `b:zipfile` is not set there might still be some problems.

For example:

* open two different zip files on different tabs
* position the cursor on some file
* click on another tab header to switch to that tab
* see that a file is automatically opened even though that file is not clicked on

I think it would be more expected if no file is opened just by switching tab with a mouse.

**Comment by brammool** [comment]
> I think even if `b:zipfile` is not set there might still be some problems.
> 
> For example:
> 
> * open two different zip files on different tabs
> * position the cursor on some file
> * click on another tab header to switch to that tab
> * see that a file is automatically opened even though that file is not clicked on
> 
> I think it would be more expected if no file is opened just by switching tab with a mouse.

True, but that is more work...  I suppose the function could check
"winid" from getmousepos() or v:mouse_winid to check if the click was
inside the window.

-- 
hundred-and-one symptoms of being an internet addict:
271. You collect hilarious signatures from all 250 mailing lists you
     are subscribed to.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by user202729** [comment]
Is there anything that relies on mapping on clicks outside the window though? We can also just not trigger `map <buffer> <leftmouse>` mappings if they're handled by vim itself (tab/window switch)?
(or trigger the mouse click mapping on the `<buffer>` mappings in the new window)

**Comment by ronin49** [comment]
what is the status of this?

**Comment by ronin49** [comment]
do not trigger mouse click mapping if mouse click was not on buffer?

**Comment by chrisbra** [comment]
It's strange, because this particular error should have been fixed by 71badf9547e8f89571b9a095183671cbb333d528

If you can reproduce, please show an updated stack trace.

> do not trigger mouse click mapping if mouse click was not on buffer?

That is already the case. The Mouse mapping is *buffer-local*

**Comment by ronin49** [comment]
no! please try the following:
1. create 2 tabs, open different files in each
2. :noremap <silent> <buffer>    <leftmouse>     <leftmouse>:echom 123<cr>
3. switch tab using mouse and see mapping will execute

**Comment by chrisbra** [comment]
Please, we have been talking in the context of the zip plugin. And the zip plugin does use a [buffer-local mapping for LeftMouse](https://github.com/vim/vim/blob/ebea31e454b9a1731cde845226f2c28ca5c097b1/runtime/autoload/zip.vim#L158). And please read `:h :map-local`

**Comment by ronin49** [comment]
yes i use terminal

**Comment by chrisbra** [comment]
well, anyhow, I cannot reprodue the issue with Vim 9.1.1000

**Comment by ronin49** [comment]
i can!
os: windows 10
vim: compiled from source using msvc,
commit 4f73c07abff420bad9fa5befc2c284c00b984993 (HEAD -> master, tag: v9.1.0998, origin/master, origin/HEAD)

**Comment by chrisbra** [comment]
I mean, please post the error message you see

**Comment by ronin49** [comment]
yes! i cannot reproduce in gvim also! so let's fix it in terminal!

**Comment by zeertzjq** [comment]
> It's strange, because this particular error should have been fixed by [71badf9](https://github.com/vim/vim/commit/71badf9547e8f89571b9a095183671cbb333d528)
> 
> If you can reproduce, please show an updated stack trace.
> 
> > do not trigger mouse click mapping if mouse click was not on buffer?
> 
> That is already the case. The Mouse mapping is _buffer-local_

No it isn't. Mappings do not depend on which buffer the mouse click is in. They only depend on which buffer the cursor is in.

**Comment by chrisbra** [comment]
Please show the error message

**Comment by ronin49** [comment]
what error?
in terminal when you click tab via mouse it is triggering mouse click mapping!

**Comment by chrisbra** [comment]
This issue here specificially about an error message in the zip plugin, when clicking with the mouse on the tabline (in terminal Vim) to load a different buffer.

I take your last message as you cannot reproduce this issue either. So I am inclined to close this issue as fixed.

**Comment by ronin49** [comment]
yes let's close

**Comment by chrisbra** [comment]
fixed as of 71badf9547e8f89571b9a095183671cbb333d528
